File: scripts/extract-woo-products.py
#!/usr/bin/env python3
"""
Extract WooCommerce products from WordPress SQL dump
"""

import re
import csv
import sys

sql_path = sys.argv[1] if len(sys.argv) > 1 else './wpress-extracted/database.sql'
output_path = sys.argv[2] if len(sys.argv) > 2 else './scripts/woo-products-extracted.csv'

print(f"Reading SQL dump from: {sql_path}")

with open(sql_path, 'r', encoding='utf-8', errors='ignore') as f:
    content = f.read()

# Find products - columns in wp_posts:
# ID, author, post_date, post_date_gmt, content, title, excerpt, status, comment_status, 
# ping_status, password, post_name (slug), to_ping, pinged, modified, modified_gmt, 
# filtered, parent, guid, menu_order, post_type, mime_type, comment_count

print("Extracting products...")
product_pattern = r"\((\d+),\d+,'[^']*','[^']*','([^']*)','([^']+)','([^']*)','(publish|draft|private)','[^']*','[^']*','[^']*','([^']+)','[^']*','[^']*','[^']*','[^']*','[^']*',\d+,'[^']*',\d+,'product',"
products = {}

for match in re.finditer(product_pattern, content):
    product_id = match.group(1)
    description = match.group(2).replace("\\n", "\n").replace("\\'", "'")[:500]  # Truncate for CSV
    title = match.group(3).replace("\\'", "'")
    short_desc = match.group(4).replace("\\'", "'")[:200]
    status = match.group(5)
    slug = match.group(6)
    
    if status == 'publish':
        products[product_id] = {
            'id': product_id,
            'name': title,
            'slug': slug,
            'description': description,
            'short_description': short_desc,
            'sku': '',
            'price': '',
            'regular_price': '',
            'sale_price': '',
            'stock': '',
            'stock_status': '',
            'weight': '',
            'categories': [],
            'images': [],
            'old_url': f'/product/{slug}/'
        }

print(f"Found {len(products)} published products")

# Extract postmeta
print("Extracting product metadata...")
meta_pattern = r"\((\d+),(\d+),'(_price|_regular_price|_sale_price|_sku|_stock|_stock_status|_weight|_thumbnail_id|_product_image_gallery)','([^']*)'\)"

for match in re.finditer(meta_pattern, content):
    post_id = match.group(2)
    meta_key = match.group(3)
    meta_value = match.group(4)
    
    if post_id in products:
        if meta_key == '_price':
            products[post_id]['price'] = meta_value
        elif meta_key == '_regular_price':
            products[post_id]['regular_price'] = meta_value
        elif meta_key == '_sale_price':
            products[post_id]['sale_price'] = meta_value
        elif meta_key == '_sku':
            products[post_id]['sku'] = meta_value
        elif meta_key == '_stock':
            products[post_id]['stock'] = meta_value
        elif meta_key == '_stock_status':
            products[post_id]['stock_status'] = meta_value
        elif meta_key == '_weight':
            products[post_id]['weight'] = meta_value
        elif meta_key == '_thumbnail_id':
            products[post_id]['images'].append(meta_value)
        elif meta_key == '_product_image_gallery':
            products[post_id]['images'].extend(meta_value.split(','))

# Extract terms (categories)
print("Extracting categories...")
terms = {}
term_pattern = r"\((\d+),'([^']+)','([^']+)',\d+\)"
for match in re.finditer(term_pattern, content):
    term_id = match.group(1)
    name = match.group(2).replace("\\'", "'")
    slug = match.group(3)
    terms[term_id] = {'name': name, 'slug': slug}

# Extract term taxonomies (to find product categories)
print("Extracting term taxonomies...")
term_taxonomies = {}
tax_pattern = r"\((\d+),(\d+),'product_cat','[^']*',(\d+),"
for match in re.finditer(tax_pattern, content):
    term_tax_id = match.group(1)
    term_id = match.group(2)
    parent = match.group(3)
    if term_id in terms:
        term_taxonomies[term_tax_id] = {
            'term_id': term_id,
            'name': terms[term_id]['name'],
            'slug': terms[term_id]['slug'],
            'parent': parent
        }

# Build category hierarchy
def get_category_path(term_tax_id, depth=0):
    if depth > 10 or term_tax_id not in term_taxonomies:
        return []
    tax = term_taxonomies[term_tax_id]
    if tax['parent'] != '0':
        # Find parent term_tax_id
        for ttid, t in term_taxonomies.items():
            if t['term_id'] == tax['parent']:
                return get_category_path(ttid, depth + 1) + [tax['name']]
    return [tax['name']]

# Extract term relationships (product -> category)
print("Extracting product-category relationships...")
rel_pattern = r"\((\d+),(\d+),\d+\)"
for match in re.finditer(rel_pattern, content):
    object_id = match.group(1)
    term_tax_id = match.group(2)
    
    if object_id in products and term_tax_id in term_taxonomies:
        cat_path = get_category_path(term_tax_id)
        if cat_path:
            products[object_id]['categories'].append(' > '.join(cat_path))

# Extract image URLs from posts (attachments)
print("Extracting image URLs...")
attachment_pattern = r"\((\d+),\d+,'[^']*','[^']*','[^']*','[^']*','[^']*','inherit',[^)]+,'([^']+)',[^,]+,'attachment','image/[^']+',\d+\)"
attachments = {}
for match in re.finditer(attachment_pattern, content):
    att_id = match.group(1)
    guid = match.group(2)
    attachments[att_id] = guid

# Update products with image URLs
for pid, product in products.items():
    image_urls = []
    for img_id in product['images']:
        if img_id in attachments:
            image_urls.append(attachments[img_id])
    product['image_urls'] = '|'.join(image_urls)
    product['categories'] = '; '.join(set(product['categories']))

# Write to CSV
print(f"Writing {len(products)} products to {output_path}")
fieldnames = ['id', 'name', 'slug', 'sku', 'price', 'regular_price', 'sale_price', 
              'stock', 'stock_status', 'weight', 'categories', 'description', 
              'short_description', 'image_urls', 'old_url']

with open(output_path, 'w', newline='', encoding='utf-8') as f:
    writer = csv.DictWriter(f, fieldnames=fieldnames, extrasaction='ignore')
    writer.writeheader()
    for product in products.values():
        writer.writerow(product)

print("Done!")

# Summary
print("\n--- Summary ---")
print(f"Total products: {len(products)}")
print(f"Products with SKU: {sum(1 for p in products.values() if p['sku'])}")
print(f"Products with price: {sum(1 for p in products.values() if p['price'])}")
print(f"Products with images: {sum(1 for p in products.values() if p['image_urls'])}")

# Category breakdown
cat_counts = {}
for p in products.values():
    for cat in p['categories'].split('; '):
        if cat:
            top_cat = cat.split(' > ')[0]
            cat_counts[top_cat] = cat_counts.get(top_cat, 0) + 1

print("\nTop-level categories:")
for cat, count in sorted(cat_counts.items(), key=lambda x: -x[1])[:15]:
    print(f"  {cat}: {count}")

File: scripts/extract-woo-products.js
#!/usr/bin/env node
/**
 * Extract WooCommerce products from WordPress SQL dump
 */

const fs = require('fs')
const path = require('path')

const sqlPath = process.argv[2] || './wpress-extracted/database.sql'
const outputPath = process.argv[3] || './scripts/woo-products.csv'

console.log(`Reading SQL dump from: ${sqlPath}`)

const sql = fs.readFileSync(sqlPath, 'utf8')

// Find all product posts
const products = []
const productMeta = {}
const terms = {}
const termRelationships = {}
const termTaxonomies = {}

// Parse posts - look for products
const postsRegex = /INSERT INTO `SERVMASK_PREFIX_posts` VALUES \(([^;]+)\);/g
let match

console.log('Parsing posts...')
while ((match = postsRegex.exec(sql)) !== null) {
  const values = match[1]
  // Split by ),( to get individual rows
  const rows = values.split('),(')
  
  for (const row of rows) {
    // Clean up the row
    const cleanRow = row.replace(/^\(/, '').replace(/\)$/, '')
    
    // Check if this is a product
    if (cleanRow.includes("'product'") && !cleanRow.includes("'product_variation'")) {
      // Parse the row - WP posts columns:
      // ID, post_author, post_date, post_date_gmt, post_content, post_title, post_excerpt,
      // post_status, comment_status, ping_status, post_password, post_name (slug), to_ping,
      // pinged, post_modified, post_modified_gmt, post_content_filtered, post_parent,
      // guid, menu_order, post_type, post_mime_type, comment_count
      
      try {
        // Extract ID (first number)
        const idMatch = cleanRow.match(/^(\d+),/)
        if (!idMatch) continue
        const id = idMatch[1]
        
        // Extract post_title - find the pattern after the dates and content
        // This is tricky with SQL escaping, let's use a simpler approach
        const titleMatch = cleanRow.match(/'([^']*?)','[^']*?','(publish|draft|private|pending)',/)
        if (titleMatch) {
          const title = titleMatch[1].replace(/\\'/g, "'")
          const status = titleMatch[2]
          
          // Extract slug (post_name)
          const slugMatch = cleanRow.match(/'(publish|draft|private|pending)','[^']*?','[^']*?','[^']*?','([^']+)',/)
          const slug = slugMatch ? slugMatch[2] : ''
          
          if (status === 'publish') {
            products.push({
              id,
              title,
              slug,
              status
            })
          }
        }
      } catch (e) {
        // Skip malformed rows
      }
    }
  }
}

console.log(`Found ${products.length} published products`)

// Parse postmeta for prices, SKUs, etc
console.log('Parsing product meta...')
const metaRegex = /INSERT INTO `SERVMASK_PREFIX_postmeta` VALUES \(([^;]+)\);/g

while ((match = metaRegex.exec(sql)) !== null) {
  const values = match[1]
  const rows = values.split('),(')
  
  for (const row of rows) {
    const cleanRow = row.replace(/^\(/, '').replace(/\)$/, '')
    
    // postmeta: meta_id, post_id, meta_key, meta_value
    const metaMatch = cleanRow.match(/(\d+),(\d+),'([^']+)','([^']*)'/)
    if (metaMatch) {
      const postId = metaMatch[2]
      const metaKey = metaMatch[3]
      const metaValue = metaMatch[4]
      
      if (!productMeta[postId]) {
        productMeta[postId] = {}
      }
      
      // Store relevant meta
      if (['_price', '_regular_price', '_sale_price', '_sku', '_stock', '_stock_status', '_weight', '_thumbnail_id'].includes(metaKey)) {
        productMeta[postId][metaKey] = metaValue
      }
    }
  }
}

// Parse terms (categories, tags)
console.log('Parsing terms...')
const termsRegex = /INSERT INTO `SERVMASK_PREFIX_terms` VALUES \(([^;]+)\);/g

while ((match = termsRegex.exec(sql)) !== null) {
  const values = match[1]
  const rows = values.split('),(')
  
  for (const row of rows) {
    const cleanRow = row.replace(/^\(/, '').replace(/\)$/, '')
    const termMatch = cleanRow.match(/(\d+),'([^']+)','([^']+)',/)
    if (termMatch) {
      terms[termMatch[1]] = {
        name: termMatch[2].replace(/\\'/g, "'"),
        slug: termMatch[3]
      }
    }
  }
}

// Parse term_taxonomy
console.log('Parsing term taxonomies...')
const taxRegex = /INSERT INTO `SERVMASK_PREFIX_term_taxonomy` VALUES \(([^;]+)\);/g

while ((match = taxRegex.exec(sql)) !== null) {
  const values = match[1]
  const rows = values.split('),(')
  
  for (const row of rows) {
    const cleanRow = row.replace(/^\(/, '').replace(/\)$/, '')
    const taxMatch = cleanRow.match(/(\d+),(\d+),'([^']+)',/)
    if (taxMatch) {
      const termTaxId = taxMatch[1]
      const termId = taxMatch[2]
      const taxonomy = taxMatch[3]
      
      if (taxonomy === 'product_cat') {
        termTaxonomies[termTaxId] = {
          termId,
          taxonomy,
          name: terms[termId]?.name || '',
          slug: terms[termId]?.slug || ''
        }
      }
    }
  }
}

// Parse term_relationships
console.log('Parsing term relationships...')
const relRegex = /INSERT INTO `SERVMASK_PREFIX_term_relationships` VALUES \(([^;]+)\);/g

while ((match = relRegex.exec(sql)) !== null) {
  const values = match[1]
  const rows = values.split('),(')
  
  for (const row of rows) {
    const cleanRow = row.replace(/^\(/, '').replace(/\)$/, '')
    const relMatch = cleanRow.match(/(\d+),(\d+),/)
    if (relMatch) {
      const objectId = relMatch[1]
      const termTaxId = relMatch[2]
      
      if (!termRelationships[objectId]) {
        termRelationships[objectId] = []
      }
      termRelationships[objectId].push(termTaxId)
    }
  }
}

// Build final product list with all data
console.log('Building product list...')
const finalProducts = products.map(p => {
  const meta = productMeta[p.id] || {}
  const categories = (termRelationships[p.id] || [])
    .map(taxId => termTaxonomies[taxId])
    .filter(t => t)
    .map(t => t.name)
    .join(' > ')
  
  return {
    id: p.id,
    name: p.title,
    slug: p.slug,
    sku: meta._sku || '',
    price: meta._price || meta._regular_price || '',
    sale_price: meta._sale_price || '',
    stock: meta._stock || '',
    stock_status: meta._stock_status || '',
    weight: meta._weight || '',
    categories: categories,
    old_url: `/product/${p.slug}/`
  }
})

// Write to CSV
console.log(`Writing ${finalProducts.length} products to ${outputPath}`)
const headers = ['id', 'name', 'slug', 'sku', 'price', 'sale_price', 'stock', 'stock_status', 'weight', 'categories', 'old_url']
const csvContent = [
  headers.join(','),
  ...finalProducts.map(p => 
    headers.map(h => {
      const val = p[h] || ''
      // Escape quotes and wrap in quotes if contains comma or quote
      if (val.includes(',') || val.includes('"') || val.includes('\n')) {
        return `"${val.replace(/"/g, '""')}"`
      }
      return val
    }).join(',')
  )
].join('\n')

fs.writeFileSync(outputPath, csvContent)
console.log('Done!')

// Also output a summary
console.log('\n--- Summary ---')
console.log(`Total products: ${finalProducts.length}`)
const categoryCounts = {}
finalProducts.forEach(p => {
  if (p.categories) {
    const cats = p.categories.split(' > ')
    cats.forEach(c => {
      categoryCounts[c] = (categoryCounts[c] || 0) + 1
    })
  }
})
console.log('\nTop categories:')
Object.entries(categoryCounts)
  .sort((a, b) => b[1] - a[1])
  .slice(0, 20)
  .forEach(([cat, count]) => console.log(`  ${cat}: ${count}`))
